NeoML/src/TraditionalML/BytePairEncoder.cpp

```diff
@@ -153,164 +148,178 @@ void CBytePairEncoder::Decode( const CArray<int>& tokenIds,
 	words.Add( currentWord );
 }
 
-int CBytePairEncoder::Size() const
+// Returns string representation of token by tokenId.
+CString CBytePairEncoder::getToken( int shiftedTokenId ) const
+{
+	NeoAssert( shiftedTokenId >= UnknownTokenId() && shiftedTokenId < Size() + UnknownTokenId() );
+
+	if( shiftedTokenId == UnknownTokenId() ) {
+		// Unknown token.
+		return UnkToken;
+	} else {
+		return tokens[shiftedTokenId - UnknownTokenId() - 1];
+	}
+}
+
+// Removes special subtokens from token.
+void CBytePairEncoder::removeSpecialTokens( CString& token, bool& hasEow, bool& hasSow ) const
+{
+	hasEow = replaceEowToken( token, params.EndOfWordToken, "" );
+	hasSow = replaceSowToken( token, params.StartOfWordToken, "" );
+}
+
+// If Eow is enabled and if 'token' ends with 'eowToken', replaces matched suffix with 'replacement'
+bool CBytePairEncoder::replaceEowToken( CString& token, const CString& eowToken, const CString& replacement ) const
+{
+	if( !UseEndOfWordToken() || token.Length() < eowToken.Length() ) {
+		return false;
+	}
+	NeoAssert( !eowToken.IsEmpty() );
+
+	const int eowPos = token.Length() - eowToken.Length();
+	if( token.CompareSubstr( eowPos, eowToken, eowToken.Length() ) == 0 ) {
+		token.StrReplace( eowPos, eowToken.Length(), replacement );
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// If Sow is enabled and if 'token' starts with 'sowToken', replaces matched prefix with 'replacement'
+bool CBytePairEncoder::replaceSowToken( CString& token, const CString& sowToken, const CString& replacement ) const
 {
-	// One extra for 'Unknown'
-	return 1 + tokens.Size();
+	if( !UseStartOfWordToken() || token.Length() < sowToken.Length() ) {
+		return false;
+	}
+	NeoAssert( !sowToken.IsEmpty() );
+
+	if( token.CompareSubstr( 0, sowToken, sowToken.Length() ) == 0 ) {
+		token.StrReplace( 0, sowToken.Length(), replacement );
+		return true;
+	} else {
+		return false;
+	}
 }
 
+static constexpr int BytePairEncoderImplVersion = 1;
+static const CString DeprecatedSowToken( "/\xFF" );

```

Comment:
_minor_ I would call it `Legacy`. IMHO `Deprecated` means that may be we will remove it in future (but in fact we won't because we _must_ support version `0` because it's used already).